xiilib/databases.py

```diff
@@ -23,6 +28,68 @@ def restart(self) -> None:
         """Restart the application."""
 
 
+class _RedisDatabaseRequiresShim:  # pylint: disable=too-few-public-methods
```

Comment:
What is the reason for needing this please? What would the change be if we just use the redis library in its current state?

Reply:
This is due to the difficulties caused by the  [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619). I have described this in other documents.

Reply:
I'm not sure I understand the link with architecture guidance behind this. I'm also not sure we should use this approach but instead use the Redis library and treat redis differently than the other databases as it's not integrated the same way. We should add these integrations the same way they would be added in a normal charm.

Reply:
Adding Redis support is a relatively straightforward task without the architecture guidance, as described in the internal case study document. It can be easily integrated into the same database module, using Redis charm library, along with other databases sharing many common procedures. However, with the architecture guidance, it must be separated into another module and treated differently from other databases, increasing the cost of implementation and maintenance.

Reply:
As noted by Gregory, we will need to change this implementation using the redis library and the normal approach. In this case we wouldn't use an observer since the event handling methods need to be on the charm class

Reply:
I can use the Redis charm library in the shim, if you prefer that way, but it would change the overall result.

Reply:
I think about this after and I don't think it will be a good idea to use the Redis library, as the current Redis library is a very primitive charm library and it's destined to be replaced by the data platform library. If we use the Redis library now, in the future when we change to data platform library, it will cause all the charms uses Redis need to update as well. And since the Redis relation is a very simple relation (host, port, user, password in unit data bag), it's very easy to implement natively in the charm. I think it would be better to wait until we got Redis support database platform library.